File: php/src/MoneyProblem/Domain/Portfolio.php
<?php

namespace MoneyProblem\Domain;

class Portfolio {

    private $bank;
    private $wallet;

    public function __construct(Bank $bank)
    {
        $this->bank = $bank;
        $this->wallet = [];
    }

    public function addMoney(Money $money){
        $newPortfolio = new Portfolio($this->getBank());
        $isFinish = false;
        if(empty($this->wallet)){
            array_push($this->wallet,$money);
            $isFinish = true;
        }
        $currency = $money->getCurrency()->getValue();
        $find = false;
        $index = 0;
        while(!$isFinish && !$find && $index < count($this->wallet)){
            if($this->wallet[$index]->getCurrency()->getValue() == $currency){
                $find = true;
                $isFinish = true;
                $newAmount = $this->wallet[$index]->getAmount() + $money->getAmount();
                $this->wallet[$index]->setAmount($newAmount);
            }
            $index++;
        }
        if(!$find && !$isFinish){
            array_push($this->wallet,$money);
        }
        
        $newPortfolio->setWallet($this->getWallet());
        return $newPortfolio;
    }

    public function sum(Currency $currency){
        $sum = 0;
        foreach ($this->wallet as $money) {
            $sum += $this->bank->convert($money, $currency);
        }
        return $sum;
    }

    public function getAmountByCurrency(Currency $currency){
        $find = false;
        $index = 0;
        $amount = 0;
        while(!$find && $index < count($this->wallet)){
            if($this->wallet[$index]->getCurrency()->getValue() == $currency){
                $find = true;
                $amount = $this->wallet[$index]->getAmount();
            }
            $index++;
        }
        return $amount;
    }


    private function getBank() : Bank {
        return $this->bank;
    }

    private function setWallet(array $wallet) : void {
        $this->wallet = $wallet;
    }

    private function getWallet() : array {
        return $this->wallet;
    }
}

File: java/src/main/java/money_problem/domain/Portfolio.java
package money_problem.domain;

import java.util.ArrayList;

public class Portfolio {
    //private ArrayList<Money> moneys;
    private ArrayList<Money> moneys;

    Portfolio(){
        moneys = new ArrayList<Money>();
    }

    public void addInPortfolio(Money myMoney){
        if(myMoney.getAmount() > 0){
            int indexCurrency = this.findCurrency(myMoney.getCurrency());
            if(indexCurrency != -1){
                moneys.get(indexCurrency).addCurrency(myMoney.getAmount());
            }else{
                moneys.add(new Money(myMoney.getCurrency(), myMoney.getAmount()));
            }
        }
    }

    public double getMoney(Currency currency){
        int indexCurrency = this.findCurrency(currency);
        if(indexCurrency != -1){
            return moneys.get(indexCurrency).getAmount();
        }else{
            return 0;
        }
    }

    public int findCurrency(Currency currency){
        int index = 0;
        int indexCurrency = -1;
        while(indexCurrency == -1 && index != this.getMoneys().size()){
            if(this.getMoneys().get(index).getCurrency() == currency){
                indexCurrency = index;
            }
            index++;
        }
        return indexCurrency;
    }

    public ArrayList<Money> getMoneys(){
        return this.moneys;
    }
}
